taskcat/s3_sync.py

```diff
@@ -0,0 +1,157 @@
+                    break
+            if not exclude_path:
+                for file in files:
+                    exclude = False
+                    for p in S3Sync._exclude_files:
+                        if fnmatch.fnmatch(file, p):
+                            exclude = True
+                            break
+                    if not exclude:
+                        full_path = root + "/" + file
+                        if include_checksums:
+                            checksum = self._hash_file(full_path)
+                        else:
+                            checksum = ""
+                        file_list.append([full_path, relpath + file, checksum])
+        return file_list
+
+    def _get_s3_file_list(self, bucket, prefix):
+        objects = []
+        more_objects = True
+        continuation_token = None
+        while more_objects:
+            if not continuation_token:
+                resp = self.s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+            else:
+                resp = self.s3_client.list_objects_v2(Bucket=bucket, Prefix=prefix, ContinuationToken=continuation_token)
+            if 'Contents' in resp:
+                objects += [[i['Key'][len(prefix):], i['ETag']] for i in resp['Contents']]
+            if 'NextContinuationToken' in resp.keys():
+                continuation_token = resp['NextContinuationToken']
+            else:
+                more_objects = False
+        return objects
+
+    def _sync(self, local_list, s3_list, bucket, prefix, acl, threads=16):
+        # determine which files to remove from S3
+        remove_from_s3 = []
+        for s3f in s3_list:
+            remove = True
+            for f in local_list:
+                if s3f[0] == f[1]:
+                    remove = False
+                    break
+            if remove:
+                remove_from_s3.append({"Key": prefix + s3f[0]})
+        # deleting objects, max 1k objects per s3 delete_objects call
+        for d in [remove_from_s3[i:i + 1000] for i in range(0, len(remove_from_s3), 1000)]:
+            self.s3_client.delete_objects(Bucket=bucket, Delete={'Objects': d})
+        # build list of files to upload
+        upload_to_s3 = []
+        for f in local_list:
+            upload = True
+            for s3f in s3_list:
+                if s3f[0] == f[1] and s3f[1] == f[2]:
+                    upload = False
+            if upload:
+                upload_to_s3.append([f[0], bucket, f[1]])
+        pool = ThreadPool(threads)
+        func = partial(self._s3_upload_file, prefix=prefix, s3_client=self.s3_client, acl=acl)
+        pool.map(func, upload_to_s3)
+        pool.close()
+        pool.join()
+
+    def _s3_upload_file(self, paths, prefix, s3_client, acl):
+        local_filename, bucket, s3_path = paths
+        retry = 0
+        while retry < 5:
+            try:
+                s3_client.upload_file(local_filename, bucket, prefix + s3_path, ExtraArgs={'ACL': acl})
```

Comment:
What's the behavior if `acl` is not provided?

Reply:
updated to default to "private"